File: scripts/double_sphere.py
import numpy as np
import math

class DoubleSphereModel:
    def __init__(self, focal_length, chi, alpha):
        self.cx = 0.5
        self.cy = 0.5
        self.focal_length = focal_length
        self.chi = chi
        self.alpha = alpha
        self.fov = self.calc_fov()

    def unproj(self, uv):
        cx = self.cx
        cy = self.cy
        fx = self.focal_length
        fy = self.focal_length
        alpha = self.alpha
        chi = self.chi

        mx = (uv[0] - cx) / fx
        my = (uv[1] - cy) / fy

        r2 = mx * mx + my * my
        beta1 = 1. - (2. * alpha - 1) * r2
        if beta1 < 0:
            return -1
        mz = (1. - alpha * alpha * r2) / (alpha * math.sqrt(beta1) + 1. - alpha)

        beta2 = mz * mz + (1. - chi * chi) * r2

        if beta2 < 0:
            return -1

        fisheye_ray = (mz * chi + math.sqrt(mz * mz + (1. - chi * chi) * r2)) / float(mz * mz + r2) * np.mat([mx, mz, -my]) - np.mat([0, chi, 0])

        return fisheye_ray

    def proj(self, xyz):
        cx = self.cx
        cy = self.cy
        fx = self.focal_length
        fy = self.focal_length
        alpha = self.alpha
        chi = self.chi

        x = xyz[0]
        y = -xyz[2]
        z = xyz[1]

        d1 = math.sqrt(x * x + y * y + z * z)
        d2 = math.sqrt(x * x + y * y + (chi * d1 + z) ** 2)
        w1 = 0
        if alpha <= 0.5 and alpha > 0:
            w1 = alpha / (1 - alpha)
        elif alpha > 0.5:
            w1 = (1 - alpha) / alpha
        w2 = (w1 + chi) / math.sqrt(2 * w1 * chi + chi ** 2 + 1)
        # if z <= -w2 * d1:
            # return -1

        resx = fx * x / (alpha * d2 + (1 - alpha) * (chi * d1 + z)) + cx
        resy = fy * y / (alpha * d2 + (1 - alpha) * (chi * d1 + z)) + cy
        if resx > 2 * self.cx or resx < 0 or resy > 2 * self.cy or resy < 0:
            return -1
        res = np.array([resx, resy])
        if alpha > 0.5:
            theta = math.pi / 2 - self.fov / 2
            if z <= math.sin(theta) * d1:
                return -1
            # r2 = ((res[0, 0] - self.cx) / fx) ** 2 + ((res[0, 1] - self.cy) / fy) ** 2
            # if r2 > 1. / (2 * alpha - 1):
                # return -1
        return res

    def calc_fov(self):
        # if self.alpha <= 0.5:
            # mx = self.cx / self.focal_length
            # r2 = mx ** 2
            # mz = (1 - self.alpha ** 2 * r2) / (self.alpha * np.sqrt(1 - (2 * self.alpha - 1) * r2) + 1 - self.alpha)
            # beta = (mz * self.chi + np.sqrt(mz ** 2 + (1 - self.chi ** 2) / r2)) / (mz ** 2 + r2)
        # else:
            # mz = (1 - self.alpha ** 2 / (2 * self.alpha - 1)) / (1 - self.alpha)
            # mx = np.sqrt(1 / (2 * self.alpha - 1))
            # beta = (mz * self.chi + np.sqrt(mz ** 2 + (1 - self.chi ** 2) / (2 * self.alpha - 1))) / (mz ** 2 + (1 / (2 * self.alpha - 1)) ** 2)
        mx = self.cx / self.focal_length
        if self.alpha > 0.5:
            r2 = min(mx ** 2, 1 / (2 * self.alpha - 1))
        else:
            r2 = mx ** 2
        mz = (1 - self.alpha ** 2 * r2) / (self.alpha * math.sqrt(1 - (2 * self.alpha - 1) * r2) + 1 - self.alpha)
        beta = (mz * self.chi + math.sqrt(mz ** 2 + (1 - self.chi ** 2) * r2)) / (mz ** 2 + r2)
        return 2 * (math.pi / 2 - math.atan2(beta * mz - self.chi, beta * mx))

File: src/camera/double_sphere.cc
#include "double_sphere.h"
#include <cmath>
#include <algorithm>

namespace omni_slam
{
namespace camera
{

DoubleSphere::DoubleSphere(const double fx, const double fy, const double cx, const double cy, const double chi, const double alpha)
    : CameraModel("DoubleSphere"),
    fx_(fx),
    fy_(fy),
    cx_(cx),
    cy_(cy),
    chi_(chi),
    alpha_(alpha)
{
    cameraMat_ << fx_, 0, cx_, 0, fy_, cy_, 0, 0, 1;
    fov_ = GetFOV();
}

bool DoubleSphere::ProjectToImage(const Vector3d &bearing, Vector2d &pixel) const
{
    double x = bearing(0);
    double y = bearing(1);
    double z = bearing(2);

    double d1 = bearing.norm();
    double d2 = sqrt(x * x + y * y + (chi_ * d1 + z) * (chi_ * d1 + z));
    double w1 = 0;
    if (alpha_ <= 0.5 && alpha_ > 0)
    {
        w1 = alpha_ / (1 - alpha_);
    }
    else if (alpha_ > 0.5)
    {
        w1 = (1 - alpha_) / alpha_;
    }
    //double w2 = (w1 + chi_) / sqrt(2 * w1 * chi_ + chi_ * chi_ + 1);
    //if (z <= -w2 * d1)
    //{
    //    return false;
    //}

    Vector3d bearing_h;
    bearing_h << x, y, (alpha_ * d2 + (1 - alpha_) * (chi_ * d1 + z));
    Vector3d pixel_h = cameraMat_ * bearing_h;
    pixel = pixel_h.hnormalized();
    if (pixel(0) > 2 * cx_ || pixel(0) < 0 || pixel(1) > 2 * cy_ || pixel(1) < 0)
    {
        return false;
    }
    if (alpha_ > 0.5)
    {
        double theta = M_PI / 2 - fov_ / 2;
        if (z <= sin(theta) * d1)
        {
            return false;
        }
    }
    return true;
}

bool DoubleSphere::UnprojectToBearing(const Vector2d &pixel, Vector3d &bearing) const
{
    double mx = (pixel(0) - cx_) / fx_;
    double my = (pixel(1) - cy_) / fy_;

    double r2 = mx * mx + my * my;
    double beta1 = 1. - (2. * alpha_ - 1) * r2;
    if (beta1 < 0)
    {
        return false;
    }
    double mz = (1. - alpha_ * alpha_ * r2) / (alpha_ * sqrt(beta1) + 1. - alpha_);

    double beta2 = mz * mz + (1. - chi_ * chi_) * r2;

    if (beta2 < 0)
    {
        return false;
    }

    bearing << mx, my, mz;

    bearing *= (mz * chi_ + sqrt(mz * mz + (1. - chi_ * chi_) * r2)) / (mz * mz + r2);
    bearing(2) -= chi_;

    return true;
}

double DoubleSphere::GetFOV()
{
    double mx = cx_ / fx_;
    double r2;
    if (alpha_ > 0.5)
    {
        r2 = std::min(mx * mx, 1 / (2 * alpha_ - 1));
    }
    else
    {
        r2 = mx * mx;
    }
    double mz = (1 - alpha_ * alpha_ * r2) / (alpha_ * sqrt(1 - (2 * alpha_ - 1) * r2) + 1 - alpha_);
    double beta = (mz * chi_ + sqrt(mz * mz + (1 - chi_ * chi_) * r2)) / (mz * mz + r2);
    return 2 * (M_PI / 2 - atan2(beta * mz - chi_, beta * mx));
}

}
}
